{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Hard coded input data\n",
    "input_data = [\n",
    "    'ATGGAACAAGAATGA',\n",
    "    'ATGATTTAAATGATCTAAATGATTTAA',\n",
    "    'CATATGATTATTTAAATCATGATTATTTAGGATATGGATATTTAGATT',\n",
    "    'ATGATTATGTAA',\n",
    "    'ATTATGTAA',\n",
    "    'ATGCGTCGT'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. Hard coded codon-to-protein dictionary (uses T instead of U since going straight from DNA)\n",
    "# Assumption: Just coded ATG as start instead of start/methionine, since no clear heuristic for determining which one to use given that not all sequences are real proteins\n",
    "\n",
    "codon_to_protein_dict = {\n",
    "    'TTT': ['F'],\n",
    "    'TTC': ['F'],\n",
    "    'TTA': ['L'],\n",
    "    'TTG': ['L'],\n",
    "    'CTT': ['L'],\n",
    "    'CTC': ['L'],\n",
    "    'CTA': ['L'],\n",
    "    'CTG': ['L'],\n",
    "    'ATT': ['I'],\n",
    "    'ATC': ['I'],\n",
    "    'ATA': ['I'],\n",
    "    'ATG': ['START'],\n",
    "    'GTT': ['V'],\n",
    "    'GTC': ['V'],\n",
    "    'GTA': ['V'],\n",
    "    'GTG': ['V'],\n",
    "    'TCT': ['S'],\n",
    "    'TCC': ['S'],\n",
    "    'TCA': ['S'],\n",
    "    'TCG': ['S'],\n",
    "    'CCT': ['P'],\n",
    "    'CCC': ['P'],\n",
    "    'CCA': ['P'],\n",
    "    'CCG': ['P'],\n",
    "    'ACT': ['T'],\n",
    "    'ACC': ['T'],\n",
    "    'ACA': ['T'],\n",
    "    'ACG': ['T'],\n",
    "    'GCT': ['A'],\n",
    "    'GCC': ['A'],\n",
    "    'GCA': ['A'],\n",
    "    'GCG': ['A'],\n",
    "    'TAT': ['Y'],\n",
    "    'TAC': ['Y'],\n",
    "    'TAA': ['STOP'],\n",
    "    'TAG': ['STOP'],\n",
    "    'CAT': ['H'],\n",
    "    'CAC': ['H'],\n",
    "    'CAA': ['Q'],\n",
    "    'CAG': ['Q'],\n",
    "    'AAT': ['N'],\n",
    "    'AAC': ['N'],\n",
    "    'AAA': ['K'],\n",
    "    'AAG': ['K'],\n",
    "    'GAT': ['D'],\n",
    "    'GAC': ['D'],\n",
    "    'GAA': ['E'],\n",
    "    'GAG': ['E'],\n",
    "    'TGT': ['C'],\n",
    "    'TGC': ['C'],\n",
    "    'TGA': ['STOP'],\n",
    "    'TGG': ['W'],\n",
    "    'CGT': ['R'],\n",
    "    'CGC': ['R'],\n",
    "    'CGA': ['R'],\n",
    "    'CGG': ['R'],\n",
    "    'AGT': ['S'],\n",
    "    'AGC': ['S'],\n",
    "    'AGA': ['R'],\n",
    "    'AGG': ['R'],\n",
    "    'GGT': ['G'],\n",
    "    'GGC': ['G'],\n",
    "    'GGA': ['G'],\n",
    "    'GGG': ['G']\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import library\n",
    "import numpy as np\n",
    "\n",
    "# Define methods\n",
    "def get_codon(input_codon):\n",
    "    '''\n",
    "    Input: single codon DNA sequence\n",
    "    Output: corresponding protein value\n",
    "    '''\n",
    "    prot_val = codon_to_protein_dict[input_codon]\n",
    "    return prot_val[0]\n",
    "\n",
    "def translate_dna(input_dna):\n",
    "    '''\n",
    "    Input: single mini-chromosome DNA sequence\n",
    "    Output: translated protein sequence\n",
    "    Does initial assumption checks before translating\n",
    "    '''\n",
    "    # Make all bases upper-case\n",
    "    temp_dna = input_dna.upper()\n",
    "    # Check if multiple of 3\n",
    "    assert len(temp_dna) % 3 == 0, 'DNA sequence is not a multiple of 3!'\n",
    "    # Get list of available codon characters\n",
    "    available_chars = set(''.join(list(codon_to_protein_dict.keys())))\n",
    "    assert all([(i in available_chars) for i in temp_dna]), 'DNA sequence has non-codon bases!'\n",
    "    \n",
    "    # Iterate through each codon\n",
    "    translated_prot = ''\n",
    "    for i in range(int(len(temp_dna)/3)):\n",
    "        codon = get_codon(temp_dna[(i*3):(i*3)+3])\n",
    "        translated_prot += codon + '-'\n",
    "    return translated_prot[:-1]\n",
    "\n",
    "def get_start_stop(input_prot_seq):\n",
    "    '''\n",
    "    Input: input protein sequence\n",
    "    Output: index position for start and stop codon\n",
    "    '''\n",
    "    prot_list = input_prot_seq.split('-')\n",
    "    in_new_prot = False\n",
    "    start_stop_list = []\n",
    "    temp_pair_list = [-1,-1]\n",
    "    # Iterate through each amino acid\n",
    "    for prot_index in range(len(prot_list)):\n",
    "        if prot_list[prot_index] == 'START':\n",
    "            temp_pair_list[0] = prot_index            \n",
    "            start_stop_list.append(temp_pair_list.copy())\n",
    "            in_new_prot = True\n",
    "        elif prot_list[prot_index] == 'STOP':\n",
    "            if in_new_prot:\n",
    "                # Has corresponding START codon\n",
    "                start_stop_list[-1][1] = prot_index\n",
    "            else:\n",
    "                # Doesn't have corresponding START codon\n",
    "                temp_pair_list[1] = prot_index\n",
    "                # Keep record of start/stop pairs\n",
    "                start_stop_list.append(temp_pair_list.copy())\n",
    "            # Reset temporary flags\n",
    "            in_new_prot = False\n",
    "            temp_pair_list = [-1,-1]\n",
    "                        \n",
    "    return start_stop_list\n",
    "        \n",
    "def check_real_protein_pair(input_ss_pair):\n",
    "    if -1 in input_ss_pair:\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "\n",
    "def get_protein_length(input_ss_pair):\n",
    "    if -1 in input_ss_pair:\n",
    "        return -1\n",
    "    else:\n",
    "        return input_ss_pair[1]-input_ss_pair[0]-1\n",
    "    \n",
    "def check_DNA_failure(input_prot_seq):\n",
    "    '''\n",
    "    Input: translated protein sequence of a mini-chromosome\n",
    "    Output: number of DNA failures\n",
    "    '''\n",
    "    # Get start/end index\n",
    "    start_stop_list = get_start_stop(input_prot_seq)\n",
    "    dna_failures = 0\n",
    "    # Iterate through start and stop pairs\n",
    "    for ss_pair in start_stop_list:\n",
    "        if not check_real_protein_pair(ss_pair):\n",
    "            dna_failures += 1\n",
    "        elif get_protein_length(ss_pair)==0:\n",
    "            dna_failures += 1\n",
    "\n",
    "    return dna_failures\n",
    "\n",
    "# 4. Compute values for DNA sequence\n",
    "def analyze_sequence(input_prot_seq):\n",
    "    '''\n",
    "    Input: translated protein sequence of a mini-chromosome\n",
    "    Output: Print outs of sequence analysis\n",
    "    '''\n",
    "    # Get protein sequence in list format\n",
    "    prot_list = input_prot_seq.split('-')\n",
    "    # Get start/end index\n",
    "    start_stop_list = get_start_stop(input_prot_seq)\n",
    "    # Get length of real encoded proteins\n",
    "    prot_length_list = [get_protein_length(i) for i in start_stop_list if check_real_protein_pair(i)]\n",
    "    # Get number of encoded proteins\n",
    "    num_encoded_prot = len(prot_length_list)\n",
    "    # Get shortest and longest encoded protein\n",
    "    if num_encoded_prot > 0 and any([i > 0 for i in prot_length_list]):\n",
    "        # 6. Get list of encoded protein sequence\n",
    "        encoded_prot_list = [\"\".join(prot_list[i[0]+1:i[1]]) for i in start_stop_list if check_real_protein_pair(i)]\n",
    "        print(f\"Test feature: encoding of a protein. This DNA encodes for {encoded_prot_list}. It has these properties:\")\n",
    "        print('Num encoded proteins: ' + str(num_encoded_prot) + ' proteins')\n",
    "        print('Shortest protein: ' + str(min(prot_length_list)) + ' amino acids')\n",
    "        print('Longest protein: ' + str(max(prot_length_list)) + ' amino acids')\n",
    "        # Get average encoded protein length\n",
    "        print('Mean protein length: ' + str(np.mean(prot_length_list)) + ' amino acids')\n",
    "    else:\n",
    "        print('No encoded protein in sequence!')\n",
    "    # Get DNA failures\n",
    "    dna_failures = check_DNA_failure(input_prot_seq)\n",
    "    print('Number of DNA failures: ' + str(dna_failures) + ' times')\n",
    "    # Get amount of non-coding DNA\n",
    "    # Assumption: Start and End codons are included in coding DNA\n",
    "    # Assumption: START and STOP codons are not coding if not encoded protein sequence in between\n",
    "    print('Amount of non-coding DNA: ' + str(3*len(prot_list)-3*sum([i+2 for i in prot_length_list if i > 0])) + ' base pairs')\n",
    "    \n",
    "# 5. Iterate through each mini-chromosome and analyze them\n",
    "def trans_trans(input_data):\n",
    "    '''\n",
    "    Input: list of DNA sequences\n",
    "    Output: Analyzed output of DNA sequences and translated proteins\n",
    "    '''\n",
    "    for test_seq in input_data:\n",
    "        print('Input DNA Sequence: ' + test_seq)\n",
    "        test_prot_seq = translate_dna(test_seq)\n",
    "        print('Raw Translated Protein Sequence: ' + test_prot_seq)\n",
    "        analyze_sequence(test_prot_seq)\n",
    "        print('-'*50)\n",
    "        print('')  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input DNA Sequence: ATGATGACGTAA\n",
      "Raw Translated Protein Sequence: START-START-T-STOP\n",
      "Test feature: encoding of a protein. This DNA encodes for ['T']. It has these properties:\n",
      "Num encoded proteins: 1 proteins\n",
      "Shortest protein: 1 amino acids\n",
      "Longest protein: 1 amino acids\n",
      "Mean protein length: 1.0 amino acids\n",
      "Number of DNA failures: 1 times\n",
      "Amount of non-coding DNA: 3 base pairs\n",
      "--------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "testing_data = ['ATGATGACGTAA']\n",
    "# testing_data = ['ATGATGACzTAA']\n",
    "# testing_data = ['ATGATGAcGTAA']\n",
    "\n",
    "trans_trans(testing_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input DNA Sequence: ATGGAACAAGAATGA\n",
      "Raw Translated Protein Sequence: START-E-Q-E-STOP\n",
      "Test feature: encoding of a protein. This DNA encodes for ['EQE']. It has these properties:\n",
      "Num encoded proteins: 1 proteins\n",
      "Shortest protein: 3 amino acids\n",
      "Longest protein: 3 amino acids\n",
      "Mean protein length: 3.0 amino acids\n",
      "Number of DNA failures: 0 times\n",
      "Amount of non-coding DNA: 0 base pairs\n",
      "--------------------------------------------------\n",
      "\n",
      "Input DNA Sequence: ATGATTTAAATGATCTAAATGATTTAA\n",
      "Raw Translated Protein Sequence: START-I-STOP-START-I-STOP-START-I-STOP\n",
      "Test feature: encoding of a protein. This DNA encodes for ['I', 'I', 'I']. It has these properties:\n",
      "Num encoded proteins: 3 proteins\n",
      "Shortest protein: 1 amino acids\n",
      "Longest protein: 1 amino acids\n",
      "Mean protein length: 1.0 amino acids\n",
      "Number of DNA failures: 0 times\n",
      "Amount of non-coding DNA: 0 base pairs\n",
      "--------------------------------------------------\n",
      "\n",
      "Input DNA Sequence: CATATGATTATTTAAATCATGATTATTTAGGATATGGATATTTAGATT\n",
      "Raw Translated Protein Sequence: H-START-I-I-STOP-I-START-I-I-STOP-D-START-D-I-STOP-I\n",
      "Test feature: encoding of a protein. This DNA encodes for ['II', 'II', 'DI']. It has these properties:\n",
      "Num encoded proteins: 3 proteins\n",
      "Shortest protein: 2 amino acids\n",
      "Longest protein: 2 amino acids\n",
      "Mean protein length: 2.0 amino acids\n",
      "Number of DNA failures: 0 times\n",
      "Amount of non-coding DNA: 12 base pairs\n",
      "--------------------------------------------------\n",
      "\n",
      "Input DNA Sequence: ATGATTATGTAA\n",
      "Raw Translated Protein Sequence: START-I-START-STOP\n",
      "No encoded protein in sequence!\n",
      "Number of DNA failures: 2 times\n",
      "Amount of non-coding DNA: 12 base pairs\n",
      "--------------------------------------------------\n",
      "\n",
      "Input DNA Sequence: ATTATGTAA\n",
      "Raw Translated Protein Sequence: I-START-STOP\n",
      "No encoded protein in sequence!\n",
      "Number of DNA failures: 1 times\n",
      "Amount of non-coding DNA: 9 base pairs\n",
      "--------------------------------------------------\n",
      "\n",
      "Input DNA Sequence: ATGCGTCGT\n",
      "Raw Translated Protein Sequence: START-R-R\n",
      "No encoded protein in sequence!\n",
      "Number of DNA failures: 1 times\n",
      "Amount of non-coding DNA: 9 base pairs\n",
      "--------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 7. Print out analysis\n",
    "trans_trans(input_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
